Applications/yOCTScanAndFindTissueSurface.m

```diff
@@ -1,44 +1,120 @@
 
 parse(p,varargin{:});
 in = p.Results;
 xRange_mm = in.xRange_mm;
 yRange_mm = in.yRange_mm;
-resolution_mm = in.resolution_mm;
+pixel_size_um = in.pixel_size_um;
+octProbeFOV_mm = in.octProbeFOV_mm;
 isVisualize = in.isVisualize;
+octProbePath = in.octProbePath;
+output_folder = in.output_folder;
+dispersionQuadraticTerm = in.dispersionQuadraticTerm;
+focusPositionInImageZpix = in.focusPositionInImageZpix;
 
 %% Scan
-% Todo
+totalStartTime = datetime;  % Capture the starting time
+volumeOutputFolder = [output_folder '/OCTVolume/'];
+disp('Please adjust the OCT focus such that it is precisely at the intersection of the tissue and the coverslip.')
+fprintf('%s Scanning Volume...\n', datestr(datetime));
+scanParameters = yOCTScanTile (...
+    volumeOutputFolder, ...
+    xRange_mm, ...
+    yRange_mm, ...
+    'octProbePath', octProbePath, ...
+    'octProbeFOV_mm', octProbeFOV_mm, ...
+    'pixelSize_um', pixel_size_um, ...
+    'v',true  ...
+    );
+
+%% Reconstruct OCT Image for Subsequent Surface Analysis
+fprintf('Generating image...\n');
+outputTiffFile = [output_folder '\surface_analysis.tiff'];
+yOCTProcessTiledScan(...
+    volumeOutputFolder, ... Input
+    {outputTiffFile},... Save only Tiff file as folder will be generated after smoothing
+    'focusPositionInImageZpix',focusPositionInImageZpix,...
+    'dispersionQuadraticTerm',dispersionQuadraticTerm,... Use default
+    'cropZAroundFocusArea', false,...
+    'interpMethod','sinc5', ...
+    'v',true);
+[logMeanAbs, dimensions, ~] = yOCTFromTif(outputTiffFile);
+disp([newline, '-- Data loaded successfully.']);
 
 %% Estimate tissue surface
-% Todo
+fprintf('%s Identifying tissue surface...\n', datestr(datetime));
+startTimeSurfaceDetection = datetime; % Start timing the surface detection
+dimensions = yOCTChangeDimensionsStructureUnits(dimensions,'millimeters'); % Make sure it's in mm
+surfacePosition_mm = yOCTFindTissueSurface(logMeanAbs, dimensions, 'isVisualize', isVisualize);
+x_mm=dimensions.x.values;
+y_mm=dimensions.y.values;
+fprintf('Surface identification completed in %s.\n', datestr(datetime - startTimeSurfaceDetection, 'HH:MM:SS'));
 
 %% Clean up
-% delete temporary files
+if exist(output_folder, 'dir')
+    rmdir(output_folder, 's'); % Remove the output directory after processing
+end
+totalEndTime = datetime;  % Capture the ending time
+totalDuration = totalEndTime - totalStartTime;
+fprintf('Total surface analysis process completed in %s.\n', datestr(totalDuration, 'HH:MM:SS'));
 
-%% Return
-% Tbd, this is just a demo
-surfacePosition_mm = ones(10,10);
-x_mm=1:10;
-y_mm=1:10;
+%% Check if the average surface distance is within the acceptable range
+average_surface_distance = nanmean(nanmean(surfacePosition_mm)); % Calculate the average surface distance
+acceptableRange_mm = 0.01; % 10 microns is acceptable range
+if isnan(average_surface_distance)
+    warning(['No tissue identification possible. Likely the tissue is out of focus below the detection range. ' ...
```

Comment:
That should be an error not a warning

Reply:
I started to do it as an error, but the script stopped execution without returning values

The reason I implemented this as a warning and not an error is to allow the script to complete and still return the average_surface_distance value.

By handling the issue through the isSurfaceInFocus Boolean, I thought we could get the function to signal the problem without stopping execution and allow the calling function to handle the issue appropriately. 

For example, I use the following assertion in Demo_ScanAndProcess_3D line 60:

assert(isSurfaceInFocus, 'Adjust the stage as needed and rerun the script.');

This way:

- The function completes execution and returns all relevant outputs.
- The caller can decide how to handle the issue, maintaining flexibility and control.

If you think an error is still more appropriate, I can adjust the design, please let me know

Reply:
Got it. Okay